src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -112,142 +74,99 @@ public static bool TryMoveDown(Project project, IProjectTree projectTree)
+            return TryMoveElements(elements, referenceElement, MoveAction.Above);
         }
 
         /// <summary>
-        /// Move a project tree below the project tree.
+        /// Move the respective item elements below the target.
         /// </summary>
-        public static Task<bool> TryMoveBelowAsync(ConfiguredProject configuredProject, IProjectTree projectTree, IProjectTree target)
+        public static bool TryMoveElementsBelow(Project project, ImmutableArray<ProjectItemElement> elements, IProjectTree target)
         {
-            Requires.NotNull(configuredProject, nameof(configuredProject));
-            Requires.NotNull(projectTree, nameof(projectTree));
+            Requires.NotNull(project, nameof(project));
             Requires.NotNull(target, nameof(target));
 
-            return TryMoveAsync(configuredProject, projectTree, target, MoveAction.Down);
+            var referenceElement = GetReferenceElement(project, target, MoveAction.Below);
+            return TryMoveElements(elements, referenceElement, MoveAction.Below);
         }
 
         /// <summary>
-        /// Gets the last child of a project tree.
-        /// The child will have a valid display order.
-        /// Returns null if there are no children, or no children with a valid display order.
+        /// Move the respective item elements to the top of the target's children.
         /// </summary>
-        public static IProjectTree GetLastChild(IProjectTree projectTree)
+        public static bool TryMoveElementsToTop(Project project, ImmutableArray<ProjectItemElement> elements, IProjectTree target)
         {
-            Requires.NotNull(projectTree, nameof(projectTree));
+            Requires.NotNull(project, nameof(project));
+            Requires.NotNull(target, nameof(target));
 
-            return GetChildren(projectTree).LastOrDefault();
-        }
+            // Get the target's first child. We use that child as our reference to move.
+            var targetChild = GetChildren(target).FirstOrDefault();
 
-        /// <summary>
-        /// Gets the first child of a project tree.
-        /// The child will have a valid display order.
-        /// Returns null if there are no children, or no children with a valid display order.
-        /// </summary>
-        public static IProjectTree GetFirstChild(IProjectTree projectTree)
-        {
-            Requires.NotNull(projectTree, nameof(projectTree));
-
-            return GetChildren(projectTree).FirstOrDefault();
-        }
+            // If we didn't find a child and our target is an empty folder and not the project root, let's walk up the tree to find a new target child.
+            // Empty folders do not have a valid display order currently in CPS. If they ever do, we have to make changes to this.
+            if (targetChild == null && target.IsFolder && !target.Flags.Contains(ProjectTreeFlags.ProjectRoot))
```

Comment:
This seems strange - if we're told to move the item to the top of an empty folder, we instead move it to the top of the parent folder?

Reply:
That's correct. The reason is I need choose the correct XML element to move the added item(s) above it. Empty folders do have an XML element, but they are not part of the display order. In the future, if we do consider empty folders to have a display order, this logic will be changed.

Reply:
I completely agree it will be surprising for users. Today is has worse behavior, this will make it a little less surprising.

If we are going to be able to order folders, we need to make a change to CPS. Once I get paste handling in, I can look at the folder work.

Reply:
`TryMoveElementsToTop` is used in the hint receiver and tests.